pkg/parser/parser.go

```diff
@@ -0,0 +1,291 @@
+type Parser interface {
+	Parse(context.Context, ...BackendOption) (*Package, error)
+}
+
+// parser is a Parser implementation with pluggable backends.
+type parser struct {
+	backend Backend
+}
+
+// New returns a new parser.
+func New(backend Backend) Parser {
+	return &parser{
+		backend: backend,
+	}
+}
+
+// BackendOption modifies the parser backend. Backends may accept options at
+// creation time, but must accept them at initialization.
+type BackendOption func(Backend)
+
+// Backend provides a source for the parser.
+type Backend interface {
+	Init(context.Context, ...BackendOption) (io.ReadCloser, error)
+}
+
+// Package includes fields that all meaningful parsers should support.
+type Package struct {
+	provider                     *pkgmetav1alpha1.Provider
+	configuration                *pkgmetav1alpha1.Configuration
+	customResourceDefinitions    map[string]*apiextensions.CustomResourceDefinition
+	compositeResourceDefinitions map[string]*apiextensionsv1alpha1.CompositeResourceDefinition
+	compositions                 map[string]*apiextensionsv1alpha1.Composition
+}
+
+// NewPackage returns a new Package with maps initialized.
+func NewPackage() *Package {
+	return &Package{
+		customResourceDefinitions:    map[string]*apiextensions.CustomResourceDefinition{},
+		compositeResourceDefinitions: map[string]*apiextensionsv1alpha1.CompositeResourceDefinition{},
+		compositions:                 map[string]*apiextensionsv1alpha1.Composition{},
+	}
+}
+
+// GetProvider gets the package provider manifest.
+func (p *Package) GetProvider() *pkgmetav1alpha1.Provider {
+	return p.provider
+}
+
+// GetConfiguration gets the package configuration manifest.
+func (p *Package) GetConfiguration() *pkgmetav1alpha1.Configuration {
+	return p.configuration
+}
+
+// GetCustomResourceDefinitions gets a package's custom resource definitions.
+func (p *Package) GetCustomResourceDefinitions() map[string]*apiextensions.CustomResourceDefinition {
+	return p.customResourceDefinitions
+}
+
+// GetCompositeResourceDefinitions gets a package's composite resource definitions.
+func (p *Package) GetCompositeResourceDefinitions() map[string]*apiextensionsv1alpha1.CompositeResourceDefinition {
+	return p.compositeResourceDefinitions
+}
+
+// GetCompositions gets a package's compositions.
+func (p *Package) GetCompositions() map[string]*apiextensionsv1alpha1.Composition {
+	return p.compositions
+}
+
+// Parse is the underlying parsing logic for parsing Crossplane packages.
+func (p *parser) Parse(ctx context.Context, opts ...BackendOption) (*Package, error) { //nolint:gocyclo
+	pkg := NewPackage()
+	reader, err := p.backend.Init(ctx, opts...)
+	if err != nil || reader == nil {
+		return pkg, err
+	}
+	defer func() { _ = reader.Close() }()
+	d := yaml.NewYAMLToJSONDecoder(reader)
+	for {
+		u := &unstructured.Unstructured{}
+		err := d.Decode(u)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return pkg, err
+		}
+		switch u.GroupVersionKind() {
```

Comment:
Could we load up a `runtime.Scheme` with all the types we support, then call its `Convert` method to convert on each of our Unstructured types, rather than needing type switches here?

Reply:
@negz just to make sure I am following here, this would only be possible if we moved to representing everything as a `runtime.Object` right? We would essentially be saying "please convert this if its a type we know about", but we wouldn't actually be checking what the type was?

Reply:
Right.

Reply:
I think that'd make this specific function simpler but we will have to do type casting whenever we'd like to use these objects. Since they are grouped by their types anyway, I'd actually vote for the current implementation where the casting is handled only once and here in one place.

Reply:
@muvaf I have actually been swayed to agree with @negz here because I think the types that consumers of this package want to identify may differ. In fact, its possible that we could have types in the `YAML` stream in the future that are not Crossplane types at all, and they are only emitted when special arguments are passed to the package entrypoint. Dealing with `runtime.Object` makes this parser agnostic about what is included in the `YAML` stream.

Reply:
> In fact, its possible that we could have types in the YAML stream in the future that are not Crossplane types at all, and they are only emitted when special arguments are passed to the package entrypoint. Dealing with runtime.Object makes this parser agnostic about what is included in the YAML stream.

I'm suspicious about that. I mean `Parse` literally returns `*Package` which has the following fields:
```golang
// Package includes fields that all meaningful parsers should support.
type Package struct {
	provider                     *pkgmetav1alpha1.Provider
	configuration                *pkgmetav1alpha1.Configuration
	customResourceDefinitions    map[string]*apiextensions.CustomResourceDefinition
	compositeResourceDefinitions map[string]*apiextensionsv1alpha1.CompositeResourceDefinition
	compositions                 map[string]*apiextensionsv1alpha1.Composition
}
```

So, they are actually grouped by types already. Having the following doesn't look very useful to me because it delegates the type casting to consumer even though it knows their types for sure:
```golang
// Package includes fields that all meaningful parsers should support.
type Package struct {
	provider                     runtime.Object
	configuration                runtime.Object
	customResourceDefinitions    map[string]runtime.Object
	compositeResourceDefinitions map[string]runtime.Object
	compositions                 map[string]runtime.Object
}
```

If the suggestion is to make it completely agnostic like:
```golang
// Package includes fields that all meaningful parsers should support.
type Package struct {
	objects                     []runtime.Object
}
```

Then I'm not sure if there is much value in thinking about the parser as something people will import at all since it's a generic for loop YAML-reader at that point that is easier to write than depend on.

Reply:
I think the proposal would be:
```
// Package includes fields that all meaningful parsers should support.
type Package struct {
        meta                        []runtime.Object
	objects                     []runtime.Object
}
```

You pass in two different schemes, one that recognizes meta types and one that represents objects that are to be created.